File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.cc
#include <iostream>

using namespace std;

struct Item{
    int weight, value;
};

int value[101][100001];
int n, k;
Item items[101]; 

int main()
{
    cin >> n >> k;
    for(auto i =1;i<=n;i++){
        int w, v;
        cin >> w >> v;
        items[i] = {w,v};
    }
    
    // knapsack dp
    for(auto i=1;i<=n;i++){
        for(auto j=1; j<=k; j++){
            int wi = items[i].weight;
            int vi = items[i].value;
            if(wi > j){ 
                // 현재 무게에 지금 선택된 것을 포함 x
                value[i][j] = value[i-1][j];
            }else{
                // 무게는 가능, 가치를 따짐.
                // 지금 선택된 것을 포함 X vs 포함 O
               value[i][j] = max(value[i-1][j], vi + value[i-1][j-wi]);
            }
            
        }
    }
    cout << value[n][k] << '\n';
    return 0;
}

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.java
import java.io.*;
import java.util.*;

public class Main {

	static class Item {
		int weight;
		int value;

		public Item(int weight, int value) {
			this.weight = weight;
			this.value = value;
		}

		@Override
		public String toString() {
			return "Item [weight=" + weight + ", value=" + value + "]";
		}

	}

	static int N, K;
	static Item[] items;
	static int[][] dp;
	static int MAX_WEIGHT = 100000;

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		K = Integer.parseInt(st.nextToken());

		items = new Item[N+1];
		dp=new int[N+1][K+1];
		
		for(int d[]: dp) {
			Arrays.fill(d, 0);
		}
		
		int w, v;
		for (int i = 1; i <= N; i++) {
			st = new StringTokenizer(br.readLine());
			w = Integer.parseInt(st.nextToken());
			v = Integer.parseInt(st.nextToken());
			items[i] = new Item(w, v);
//			System.out.println(items[i].toString());
		}
		
		for(int i=1; i<=N;i++) {
			for(int j=1; j<=K; j++) {
				if(j>=items[i].weight) {
					dp[i][j] = Math.max(dp[i-1][j], items[i].value+dp[i-1][j-items[i].weight]);
					continue;
				}
				
				// 현재 물건을 담을 수 없는 무게일 때는 이전 무게 그대로 내려오기
				dp[i][j]=dp[i-1][j];
			}
		}
//		System.out.println(Arrays.deepToString(dp));
		System.out.println(dp[N][K]);
		
	}

}
